### PR TITLE
add PageView cacheExtent

### PR DESCRIPTION
**Pull request by lukepighetti**
## Description

Non-breaking change that allows PageView to specify cacheExtents in whole pages. Ie, setting cacheExtent to 1 will build the current page and the one before and the one after it.

## Related Issues

https://github.com/flutter/flutter/issues/31191

## Tests

I added the following tests:

```
'cacheExtent 0'
'cacheExtent 1'
'cacheExtent 2'
```

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by lukepighetti** [comment]
Did this PR on master which appears to be a little dirty, but my PR & test coverage is clean.

File: packages/flutter/lib/src/widgets/page_view.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/page_view.dart]: cacheExtent is a bit misleading. It is not unit of pixel but of page. Maybe cachePageExtent or cachePage?

**Review by chunhtai — COMMENTED** [review]
The analyzer seems to be unhappy :(

**Review by goderbauer — CHANGES_REQUESTED** [review]
Before this can be merged we need to double-check that it doesn't break the accessibility of the PageView especially on iOS. On iOS the engine uses the cache extent as a signal for accessibility scrolling.

With voiceover turned on, the user **should not be able to** scroll to the next page just by advancing the accessibility focus by swing left or right.

Technically, we have the hasImplicitScrolling flag [1] to control that, but it looks like that's currently only wired up on Android and not on iOS [2]. We probably need to change that in order to allow PageView to have a cacheExtent.

I am marking this as "request changes" until it has either been verified that this does not break accessibility on iOS, or (if it does) until that has been fixed.

[1] https://master-api.flutter.dev/flutter/dart-ui/SemanticsFlag/hasImplicitScrolling-constant.html
[2] https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm

File: packages/flutter/lib/src/widgets/page_view.dart
Comment (lukepighetti) [reply packages/flutter/lib/src/widgets/page_view.dart]: done (cachePageExtent)

File: packages/flutter/lib/src/widgets/page_view.dart
Comment (lukepighetti) [reply packages/flutter/lib/src/widgets/page_view.dart]: done

File: packages/flutter/lib/src/widgets/page_view.dart
Comment (lukepighetti) [reply packages/flutter/lib/src/widgets/page_view.dart]: done

File: packages/flutter/lib/src/widgets/page_view.dart
Comment (lukepighetti) [reply packages/flutter/lib/src/widgets/page_view.dart]: done

File: packages/flutter/lib/src/widgets/page_view.dart
Comment (lukepighetti) [reply packages/flutter/lib/src/widgets/page_view.dart]: Not sure what I can do about having forked from master when the build is broken. Any guidance is appreciated on how to handle this in the future.

**Comment by lukepighetti** [comment]
@goderbauer does the `flutter_test` framework have any way to lock that behavior down?

**Comment by goderbauer** [comment]
flutter_test is for unit tests. This would require an integration test that verifies Flutter's interaction with iOS VoiceOver.

**Comment by lukepighetti** [comment]
@goderbauer what do you need from me at this point for this to land?

**Comment by goderbauer** [comment]
I just verified: Setting a cacheExtend on a pageView breaks accessibility scrolling on iOS. Before we can land this, we need to change the iOS accessibility bridge in the engine [1] to ensure that it properly respects the `SemanticsFlag.hasImplicitScrolling`. If that flag is set, implicit scrolling should not be enabled even if a cache extent is specified. 

[1] https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm

**Comment by lukepighetti** [comment]
@goderbauer Thanks for the follow up. I'll assume there's nothing I can do to assist until that is complete.

**Comment by goderbauer** [comment]
If you want, you could make that change to the accessibility bridge. Other than that there's not much to do.

**Comment by lukepighetti** [comment]
@goderbauer It's an area I'm not comfortable with, it would need some serious code review. Is this something that is coming down the pipeline or is it only to be implemented with this PR?

**Comment by goderbauer** [comment]
@lukepighetti The work for that is currently not planned, but we need to do it at some point.

/cc @darrenaustin Maybe you're interested in this?

I am going to close this PR for now as we cannot merge it without the engine side changes and nobody is currently working on those.

**Comment by aytunch** [comment]
Why is this closed? We need this pr

**Comment by rafay-tariq** [comment]
> 'cacheExtent 0'
> 'cacheExtent 1'
> 'cacheExtent 2'

where to add these lines ? kindly tell me.  i added inside the page builder as well as in page view but not works,
_pageController = PageController(
      initialPage: 0,
      keepPage: true,
      viewportFraction: 1.0,     
      
    );

**Comment by nocduro** [comment]
I just stumbled upon this and it looks like there was a recent pull request that potentially added the required accessibility support on iOS that was blocking this: https://github.com/flutter/engine/pull/26671

#31191

**Comment by chunhtai** [comment]
@nocduro The accessibility scroll support in the PR is tangential to the problem @goderbauer mentioned

**Comment by LuChenyang3842** [comment]
The flutter team should somehow provide the preload ability for pageview..... can someone tell me when this feature will be planned.

**Comment by Bonsai11** [comment]
Absolutely need this aswell, as I have a form for data entry spread across several pages. I know that there is a package out there to enable this, but I always prefer to have a "native" solution. With the current PageView, after you scrolled through all the none preloaded pages and then back to the beginning, isn't the situation you are now in exactly the same as if they were preloaded from the beginning? So this simply seems to be a shortcut to get somewhere where you can get already anyway, but also enabling more flexability in how you use the PageView.

**Pull request closed without merging**
